Plenigo 874 add new app interfaces to php

SebaSOFT opened the pull request:
Done with this...

Commit:
- First three methods for the API

Commit:
- hasUserBought method

Commit:
- Delete method
- leverage cURL's post options
- Some PHPDocs

Commit:
- rest of the methods
- mobile secret and all
- added delete mothods and json posting

Commit:
- Some tests

Commit:
- Fixed the associative array of the apps
- Added Tests

Commit:
- AppManagement Tests completed

maximilianschweitzer commented on src/plenigo/internal/ApiURLs.php:
```diff
@@ -84,4 +84,34 @@
+    const GET_PROD_ACCESS = "/api/v1/access/app/{USER_ID}/{PROD_ID}/{APP_ID}";
+
+    /**
+     * This URL is used to delete a 3rd party access from a certain product
```
Not for a third party. This is the merchant itself.

SebaSOFT replied:
done

maximilianschweitzer commented on src/plenigo/internal/ApiURLs.php:
```diff
@@ -84,4 +84,34 @@
+    /**
+     * This URL is used to delete a 3rd party access from a certain product
+     */   
+    const GET_DELETE_APP = "/api/v1/access/app/{USER_ID}/{APP_ID}";
```
why get_delete?

SebaSOFT replied:
Done

maximilianschweitzer commented on src/plenigo/internal/ApiURLs.php:
```diff
@@ -84,4 +84,34 @@
+    const GET_APP_TOKEN = "/api/v1/access/app/token/{USER_ID}";
+
+    /**
+     * This URL is used for 3rd party apps to get all customer apps
```
Not for a third party. This is the merchant itself.

SebaSOFT replied:
Done

maximilianschweitzer commented on src/plenigo/internal/ApiURLs.php:
```diff
@@ -84,4 +84,34 @@
     const PAYWALL_STATE = "/api/v1/paywall/state";
+    
+    /**
+     * This URL is used for 3rd party apps to get access from a customer to a product
```
Not for a third party. This is the merchant itself.

SebaSOFT replied:
Done

maximilianschweitzer commented on src/plenigo/internal/services/Service.php:
```diff
@@ -62,6 +62,19 @@ protected static function getRequest($endPoint, $oauth = false, array $params =
+    
+    protected static function deleteRequest($endPoint, $oauth = false, array $params = array())
+    {
+        if($oauth){
```
same code is used in postJSONRequest.

maximilianschweitzer commented on src/plenigo/internal/services/Service.php:
```diff
@@ -147,7 +183,8 @@ protected function checkForErrors($response)
 
-        if ($statusCode != 200) {
+        // All 200 codes are good answers
+        if ($statusCode < 200 || $statusCode >= 300) {
```
300 is also a good answer?

SebaSOFT replied:
This if is for NO GOOD answers, so anything less than 200 and more or equal 300 is a BAD answer

maximilianschweitzer commented on src/plenigo/internal/utils/RestClient.php:
```diff
@@ -71,6 +68,28 @@ public static function get($url, array $params = array())
+            // taking out the brackets because we need to use the very same variable name
+            $queryString = preg_replace('/%5B(?:[0-9]|[1-9][0-9]+)%5D=/', '=', $query);
+
+            $url .= '?' . $queryString;
```
Is there the possibility that the URL is already having parameters?

SebaSOFT replied:
I tought so too, not happening right now but will prevent that from happening

maximilianschweitzer commented on src/plenigo/internal/utils/RestClient.php:
```diff
@@ -71,6 +68,28 @@ public static function get($url, array $params = array())
+            $url .= '?' . $queryString;
+        }
+        $clazz = get_class();
+        PlenigoManager::notice($clazz, "GET DELETE URL CALL=" . $url);
```
GET DELETE?

SebaSOFT replied:
Done

maximilianschweitzer commented on src/plenigo/models/AppAccessData.php:
```diff
@@ -0,0 +1,117 @@
+     * Maps an associative array to the AppAccessData object
+     * 
+     * @param array $map an associative array representing the AppAccessData object
+     * @return \plenigo\models\AppAccessData
```
A unified formatting would be great. So our defaults are an empty line betweeen param and return.

SebaSOFT replied:
Cant customize that in BetBeans
http://stackoverflow.com/questions/5922828/how-do-i-customize-netbeans-php-auto-docblock-for-methods-and-attributes

Should I switch to Eclipse? Because the formatter cant enforce that either...

maximilianschweitzer replied:
Don't think that is worth the change. Only wondered about the inconsistency.

maximilianschweitzer commented on src/plenigo/models/MobileSecretData.php:
```diff
@@ -0,0 +1,93 @@
+class MobileSecretData {
+
+    /**
+     *
```
Missing comment?

SebaSOFT replied:
Done

maximilianschweitzer commented on src/plenigo/models/MobileSecretData.php:
```diff
@@ -0,0 +1,93 @@
+    private $email;
+
+    /**
+     *
```
Missing comment?

SebaSOFT replied:
Done

maximilianschweitzer commented on src/plenigo/services/AppManagementService.php:
```diff
@@ -0,0 +1,262 @@
+     * @param string $customerId the Customer ID to send to the API
+     * @param string $productId  the Product ID to send to the API
+     * @param string $description the App Access Description to send to the API
+     * @return AppTokenData the access token to get the App ID
```
Formatting?

SebaSOFT replied:
Manually done

maximilianschweitzer commented on src/plenigo/services/AppManagementService.php:
```diff
@@ -0,0 +1,262 @@
+        try {
+            $data = $appTokenRequest->execute();
+        } catch (Exception $exc) {
+            $errorCode = ErrorCode::getTranslation(ApiURLs::GET_APPS_ID, $exc->getCode());
```
Same code as above. This code is repeated in every method.

SebaSOFT replied:
it is not always the same, but Im extracting the code

maximilianschweitzer commented on src/plenigo/services/MobileService.php:
```diff
@@ -0,0 +1,212 @@
+     *
+     * @return The request's response.
+     *
+     * @throws \plenigo\PlenigoException on request error.
```
Switching between \plenigo\PlenigoException and only PlenigoException

SebaSOFT replied:
Done

rtorres87 commented on src/plenigo/models/AppAccessData.php:
```diff
@@ -0,0 +1,117 @@
+     * @param array $map an associative array representing the AppAccessData object
+     * @return \plenigo\models\AppAccessData
+     */
+    public static function createFromMap($map) {
```
missing methods documentation

SebaSOFT replied:
Done

Commit:
- Comments corrections...
- Execution abstraction
- Now 0% repeated code...phpcpd says

Commit:
- Comments and formatting

Commit:
- All Tests done
- Fixed a couple of problems with object/array convertion

maximilianschweitzer commented on src/plenigo/internal/utils/RestClient.php:
```diff
@@ -71,6 +68,32 @@ public static function get($url, array $params = array())
+        return new static($curlRequest);
+    }
+
+    public static function delete($url, array $params = array()) {
```
Missing comment

SebaSOFT replied:
Fixed

maximilianschweitzer commented on tests/plenigo/services/AppManagementServiceMock.php:
```diff
@@ -0,0 +1,34 @@
+
+    public static $requestResponse;
+
+    public function __construct($request)
```
comment

Commit:
- Fixed missing comment

maximilianschweitzer commented on src/plenigo/services/MobileService.php:
```diff
@@ -0,0 +1,176 @@
+     * @return MobileSecretData the email and secret for the customer mobile
+     * @throws PlenigoException
+     */
+    public static function createMobileSecret($customerId) {
```
What I'm missing here is the size parameter. Because the merchant can optionally decide how long the mobile secret will be.

Commit:
- size parameter

maximilianschweitzer commented on src/plenigo/services/MobileService.php:
```diff
@@ -0,0 +1,186 @@
+     * @throws PlenigoException
+     */
+    public static function createMobileSecret($customerId, $size) {
+        if ($size < 8) {
```
6 would be good as lower value. And also a constant for that would be useful ;-)

SebaSOFT replied:
Done

Commit:
- Cleaning up some mpty responses
- Better handling 200 responses
- Better handling of empty string responses assumed as JSON objects
